gapic/src/main/com/google/gapid/perfetto/models/CounterTrack.java

```diff
@@ -158,44 +167,46 @@ private String rangeSql(TimeSpan ts) {
   public static class Data extends Track.Data {
     public final long[] ts;
     public final double[] values;
+    public final long[] ids;
 
-    public Data(DataRequest request, long[] ts, double[] values) {
+    public Data(DataRequest request, long[] ts, double[] values, long[] ids) {
       super(request);
       this.ts = ts;
       this.values = values;
+      this.ids = ids;
     }
 
     public static Data empty(DataRequest req) {
-      return new Data(req, new long[0], new double[0]);
+      return new Data(req, new long[0], new double[0], new long[0]);
     }
   }
 
-  public static class Values implements Selection<Values.Key>, Selection.Builder<Values> {
+  public static class Values implements Selection<Long>, Selection.Builder<Values> {
     public final long[] ts;
     public final String[] names;
     public final double[][] values;
-    private final Set<Values.Key> valueKeys = Sets.newHashSet();
+    public final long[][] ids;
+    private final Set<Long> valueKeys = Sets.newHashSet();
 
     public Values(String name, Data data) {
       this.ts = data.ts;
       this.names = new String[] { name };
       this.values = new double[][] { data.values };
+      this.ids = new long[][] { data.ids };
       initKeys();
     }
 
-    private Values(long[] ts, String[] names, double[][] values) {
+    private Values(long[] ts, String[] names, double[][] values, long[][] ids) {
       this.ts = ts;
       this.names = names;
       this.values = values;
+      this.ids = ids;
       initKeys();
     }
 
     private void initKeys() {
-      for (String name : names) {
-        // Skip the last dummy entry for the range end.
-        Arrays.stream(ts, 0, ts.length - 1)
-            .boxed()
-            .forEach(t -> valueKeys.add(new Values.Key(name, t)));
+      for (long[] keys : ids) {
+        valueKeys.addAll(Arrays.stream(keys).boxed().collect(Collectors.toSet()));
```

Comment:
I think java.util.Collections.addAll would work and that would be best if it does.
If it doesn't, then you actually want to use forEach like before.  Using collect will unnecessary create a Set to contain the value of the stream.
Also, I don't think you need to boxed it if you are using forEach, but not 100% sure.

Reply:
Rewrote it to avoid set creation.